### PR TITLE
[Merged by Bors] - feat(CategoryTheory): a family is effective epi iff its precomposition with a family of split epis is

### PR DESCRIPTION
**Pull request by dagurtomas**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/EffectiveEpi/Comp.lean [hunk 1]
```diff
@@ -19,37 +19,106 @@ open Limits
...
-def effectiveEpiStructCompOfEffectiveEpiSplitEpi {B X Y : C} (f : X ⟶ B) (g : Y ⟶ X) [IsSplitEpi g]
-    [EffectiveEpi f] : EffectiveEpiStruct (g ≫ f) :=
-  effectiveEpiStructCompOfEffectiveEpiSplitEpi' f g
-    (IsSplitEpi.exists_splitEpi (f := g)).some.section_
-    (IsSplitEpi.exists_splitEpi (f := g)).some.id
+def effectiveEpiFamilyStructCompOfEffectiveEpiSplitEpi {α : Type*} {B : C} {X Y : α → C}
+    (f : (a : α) → X a ⟶ B) (g : (a : α) → Y a ⟶ X a) [∀ a, IsSplitEpi (g a)]
+    [EffectiveEpiFamily _ f] : EffectiveEpiFamilyStruct _ (fun a ↦ g a ≫ f a) :=
+  effectiveEpiFamilyStructCompOfEffectiveEpiSplitEpi' f g
+    (fun a ↦ (IsSplitEpi.exists_splitEpi (f := g a)).some.section_)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/EffectiveEpi/Comp.lean:55]: Here (and below), you should use the `IsSplitEpi` API: `section_` and the lemma `IsSplitEpi.id`.

**Comment by joelriou** [comment]
I think we should prove the more general result using epi instead of split epi. Here is a chunk of code for the case of effective epi:
```lean
noncomputable def effectiveEpiStructOfComp {C : Type*} [Category C]
  {Z Y X : C} (g : Z ⟶ Y) (f : Y ⟶ X) [EffectiveEpi (g ≫ f)] [Epi g] :
    EffectiveEpiStruct f where
  desc {W} φ h := EffectiveEpi.desc (g ≫ f) (g ≫ φ) (fun {T} g₁ g₂ eq => by
      simpa only [assoc] using h (g₁ ≫ g) (g₂ ≫ g) (by simpa only [assoc] using eq))
  fac {W} φ h := by rw [← cancel_epi g, ← assoc, EffectiveEpi.fac]
  uniq {W} φ h m hm := EffectiveEpi.uniq (g ≫ f) _ _ _ (by rw [assoc, hm])
```

The proofs are even easier! It should be similar for epimorphic families of morphisms:

```lean
noncomputable def effectiveEpiFamilyStructOfComp {C : Type*} [Category C]
    {I : Type*} {Z Y : I → C} {X : C} (g : ∀ i, Z i ⟶ Y i) (f : ∀ i, Y i ⟶ X)
    [EffectiveEpiFamily _ (fun i => g i ≫ f i)] [∀ i, Epi (g i)] :
    EffectiveEpiFamilyStruct _ f where
  desc {W} φ h := EffectiveEpiFamily.desc _ (fun i => g i ≫ f i)
    (fun i => g i ≫ φ i) (fun {T} i₁ i₂ g₁ g₂ eq =>
      by simpa [assoc] using h i₁ i₂ (g₁ ≫ g i₁) (g₂ ≫ g i₂) (by simpa [assoc] using eq))
  fac := sorry
  uniq := sorry
```

**Comment by dagurtomas** [comment]
> I think we should prove the more general result using epi instead of split epi. Here is a chunk of code for the case of effective epi:

Of course, I didn't realise this! I'll change it. For the other direction, one needs split though, or some nice properties of the category.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**